### PR TITLE
[SPARK-9281] [SQL] use decimal or double when parsing SQL

### PR DESCRIPTION
**Pull request by davies**
Right now, we use double to parse all the float number in SQL. When it's used in expression together with DecimalType, it will turn the decimal into double as well. Also it will loss some precision when using double.

This PR change to parse float number to decimal or double, based on it's  using scientific notation or not, see https://msdn.microsoft.com/en-us/library/ms179899.aspx

This is a break change, should we doc it somewhere?

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SqlParser.scala [hunk 1]
```diff
@@ -411,6 +410,17 @@ class SqlParser extends AbstractSparkSQLParser with DataTypeParser {
     }
   }
 
+  private def toDecimalOrDouble(value: String): Any = {
+    val decimal = BigDecimal(value)
+    // follow the behavior in MS SQL Server
+    // https://msdn.microsoft.com/en-us/library/ms179899.aspx
+    if (value.contains('E') || value.contains('e)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SqlParser.scala]: typo `'e`?

**Comment by davies** [comment]
cc @yhuai @rxin

**Comment by rxin** [comment]
We should add "releasenotes" label to the JIRA ticket.

**Comment by rxin** [comment]
cc @cloud-fan for review too

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 2]
```diff
@@ -543,7 +557,7 @@ object HiveTypeCoercion {
       // compatible with every child column.
       case c @ Coalesce(es) if es.map(_.dataType).distinct.size > 1 =>
         val types = es.map(_.dataType)
-        findTightestCommonTypeAndPromoteToString(types) match {
+        findWiderCommonType(types) match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: should we apply this to `CreateArray` too?

**Comment by cloud-fan** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala (hunk 2 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: For `CreateArray`, it's good to make it strict. We can loose the rule later.

**Comment by rxin** [comment]
@davies  do we need to change the hive ql part?

**Comment by davies** [comment]
@rxin HiveQl already does this, it always parse float number as decimal.

**Comment by rxin** [comment]
Great - merging this in. Thanks.

**Comment by scwf** [comment]
hi @davies  seems this is not compatible with hiveql, HiveQl still parse float number as double. https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala#L1670-L1689

Or i missed sth?

**Comment by davies** [comment]
Yes, If you use HiveContext, you will be compatible with Hive, but SQLContext will not.

**Pull request closed without merging**